### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by glencoe**
A lot of refactoring. Most prominently using the vhdl formal grammar as a base to construct components, entities, generics and ports.

### CODE REVIEW

**Comment by glencoe** [comment]
before merging back we might want to get rid of the file creation in the other tests as well.

File: elasticai/creator/vhdl/generator/functions/generate_lstm_cell_vhd.py [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+component_name = "lstm_cell"
+file_name = component_name + ".vhd"
+architecture_name = "lstm_cell_rtl"
+
+DATA_WIDTH = 16
+FRAC_WIDTH = 8
+
+
+def main():
```
Comment (Annika10) [inline elasticai/creator/vhdl/generator/functions/generate_lstm_cell_vhd.py:28]: if possible make it more similar to the sigmoid/tanh generation

File: elasticai/creator/vhdl/language.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+    Optional,
+)
+from collections import Sequence
+
+Identifier = str
+Code = Iterable[str]
+CodeGenerator = Callable[[], Code]
+
+
+class Keywords(Enum):
```
Comment (Annika10) [inline elasticai/creator/vhdl/language.py:26]: I think the idea is nice but I wouldn't say I get everything in this implementation. A documentation would be nice, also about the usage

File: elasticai/creator/systemTests/models_definition.py [hunk 3]
```diff
@@ -138,19 +128,27 @@ def crete_brevitas_model():
...
     return model
 
 
-weigth_layers = (QConv1d, QLinear, nn.BatchNorm1d, bnn.QuantConv1d, bnn.QuantLinear)
+weight_layers = (QConv1d, QLinear, nn.BatchNorm1d, bnn.QuantConv1d, bnn.QuantLinear)
 
 
 def define_weight(layers):
+    # TODO: In case we're using this function just to have deterministic weight value generation
+    #  we should reconsider our approach here
```
Comment (Annika10) [inline elasticai/creator/systemTests/models_definition.py:143]: do you have a different idea in mind?

File: elasticai/creator/brevitas/translation_functions/translation_function_tools.py [hunk 4]
```diff
@@ -13,11 +13,11 @@ def set_quantizers(layer: Module, args: Dict) -> None:
         layer (Layer): current layer
         args (Dict): The layer args
     """
-    if isinstance(layer.quantizer, Binarize):
+    if isinstance(layer.parametrizations.weight[0], Binarize):
```
[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/brevitas/translation_functions/translation_function_tools.py:16]: yes, in short the advantage of supporting parametrization is that we get quantized weights for free

File: elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+component_name = "mac_async"
+file_name = component_name + ".vhd"
+architecture_name = "mac_async_rtl"
+
+DATA_WIDTH = 16
+DATA_FRAC = 8
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py:22]: yes... i just did not have the time to apply the change to the rest of the stuff as well... I'd like you folks to continue with the refactoring process there.

File: elasticai/creator/vhdl/generator/generator_functions.py [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+from typing import Dict, List, Callable, Any, Union, Sequence
+
+from elasticai.creator.vhdl.number_representations import (
+    FloatToSignedFixedPointConverter,
+    two_complements_representation,
+    FloatToBinaryFixedPointStringConverter,
+)
+
+
+def vhdl_add_assignment(code: list, line_id: str, value: str, comment=None) -> None:
```
[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/vhdl/generator/generator_functions.py:17]: I agree

File: elasticai/creator/systemTests/models_definition.py (hunk 3 above)
Comment (glencoe) [reply elasticai/creator/systemTests/models_definition.py:143]: couldn't we just set the random seeds accordingly?

File: elasticai/creator/vhdl/language.py (hunk 2 above)
Comment (glencoe) [reply elasticai/creator/vhdl/language.py:26]: allright, i'll try to improve it

File: elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Annika10) [reply elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py:22]: okay so should we first merge it and add it afterwards or already add it in this pull request?

File: elasticai/creator/vhdl/language.py (hunk 2 above)
Comment (glencoe) [reply elasticai/creator/vhdl/language.py:26]: hmmm... to be honest i don't really know what to add here. The implementation tries to stick very close to the vhdl formal grammar, so i don't think it makes a lot of sense without knowing or referring to the grammar.

Comment (Annika10) [reply elasticai/creator/vhdl/language.py:26]: Not for now, I will look at it tomorrow

Comment (glencoe) [reply elasticai/creator/vhdl/language.py:26]: we could also just talk about it in a quick meeting at some point

**Comment by glencoe** [comment]
@Annika10  I'd say we can merge this when you're happy with the documentation, ie. as soon as you think the core idea is clear enough so that others can continue working with it.

File: elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AhmadHusein5853) [reply elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py:22]: I am working on it , and on the lstm generation

[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py:22]: cool, but as @Annika10 already indicated i'd like to merge this one here first. I'll go ahead and merge it and open additional issues for the rest of the todos.

[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/vhdl/generator/functions/generate_mac_async_vhd.py:22]: @AhmadHusein5853 please pick one of #15 or #16 for now

File: elasticai/creator/tests/test_jit.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+import torch
+from torch import jit, threshold
+from torch.nn.functional import hardtanh
+
+from elasticai.creator.layers import Binarize
+from tensor_test_case import TensorTestCase
+
+
+class JitTests(TensorTestCase):
```
[the remark this answers is not in the document]
Comment (glencoe) [reply elasticai/creator/tests/test_jit.py:14]: I think it's fine for now, but we should discuss it in the next meeting.

File: elasticai/creator/vhdl/language.py (hunk 2 above)
Comment (glencoe) [reply elasticai/creator/vhdl/language.py:29]: rresolve with reference to #20
